controllers/cloudstackfailuredomain_controller.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2"
+	csCtrlrUtils "sigs.k8s.io/cluster-api-provider-cloudstack/controllers/utils"
+)
+
+// CloudStackFailureDomainReconciler is the k8s controller manager's interface to reconcile a CloudStackFailureDomain.
+// This is primarily to adapt to k8s.
+type CloudStackFailureDomainReconciler struct {
+	csCtrlrUtils.ReconcilerBase
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackfailuredomains,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackfailuredomains/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackfailuredomains/finalizers,verbs=update
+
+// CloudStackFailureDomainReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackFailureDomains.
+// The runner does the actual reconciliation.
+type CloudStackFailureDomainReconciliationRunner struct {
+	*csCtrlrUtils.ReconciliationRunner
+	ReconciliationSubject *infrav1.CloudStackFailureDomain
+	IsoNet                *infrav1.CloudStackIsolatedNetwork
+}
+
+// Initialize a new CloudStackFailureDomain reconciliation runner with concrete types and initialized member fields.
+func NewCSFailureDomainReconciliationRunner() *CloudStackFailureDomainReconciliationRunner {
+	// Set concrete type and init pointers.
+	r := &CloudStackFailureDomainReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackFailureDomain{}}
+	r.IsoNet = &infrav1.CloudStackIsolatedNetwork{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	r.ReconciliationRunner = csCtrlrUtils.NewRunner(r, r.ReconciliationSubject, "CloudStackFailureDomain")
+
+	return r
+}
+
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CloudStackFailureDomainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	return NewCSFailureDomainReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
```

Comment:
when does `CloudStackFailureDomainReconciliationRunner.Reconcile()` get called? I wasn't quite able to trace that

Reply:
It's a part of the base stages. All reconcilers run the same base stages and once the reconciliation subject is fetched, the deletion timestamp is checked. If the deletion timestamp is present, then ReconcileDelete is picked, otherwise just Reconcile.